Improvement to the ai module

reflog opened the pull request:
Hi @abenz1267 

I've 'generalized' the ai module and created another provider - Gemini.
What do you think of this approach?

abenz1267 commented:
Hi,

thanks for working on that. I think the general approach is fine, this PR doesn't function yet though.

Small nitpicks:
- `Anthropic API returned status code 200` => avoid console messages
- `antropic.go` => `anthropic.go`
- the label should say which model/company is being used, put that info in the sub, where the model is displayed

As for not working: console spits out `Anthropic API returned status code 200`, but the answer isn't shown.

reflog commented:
@abenz1267 thanks. I've addressed your comments.
The only thing that is not clear to me is the 'sub' usage. When I put model info in it, it doesn't appear in the UI (in the screenshot I have gemini and claude with the same assistant name, but they are indistinguishable from each other)
![image](https://github.com/user-attachments/assets/c54038be-ea29-43a5-a446-975284ac8c83)

abenz1267 commented:
> @abenz1267 thanks. I've addressed your comments. The only thing that is not clear to me is the 'sub' usage. When I put model info in it, it doesn't appear in the UI (in the screenshot I have gemini and claude with the same assistant name, but they are indistinguishable from each other) ![image](https://private-user-images.githubusercontent.com/109876/423396198-c54038be-ea29-43a5-a446-975284ac8c83.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMDI2MjcsIm5iZiI6MTc0MjIwMjMyNywicGF0aCI6Ii8xMDk4NzYvNDIzMzk2MTk4LWM1NDAzOGJlLWVhMjktNDNhNS1hNDQ2LTk3NTI4NGFjOGM4My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxN1QwOTA1MjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNjNlZWFjMzRhMGM5MDgzM2QwYmM2NDFhNTQwNDJmMzg1MmQ5ZTQ2NjEwNmNjNmMxZWZhM2UxMWU0MWIyYWE1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.3LohKZdYB4TVYoN7dZU2eLLH72vl85GjFiMY2isClpA)

Modules/Plugins have a `show_sub_when_single` boolean setting that can be set to `true`. By default the `sub` won't be shown, when there's only 1 modules, f.e. when doing `walker -m ai`.

This could be simply changed in the default config.

reflog commented:
@abenz1267 lovely, that works great. I've updated the default on the ai module

abenz1267 commented on internal/modules/ai/ai.go:
```diff
@@ -0,0 +1,180 @@
+func (ai *AI) Setup() bool {
+	ai.config = config.Cfg.Builtins.AI
+	ai.provider = make(map[string]providers.Provider)
+	gemini := providers.NewGeminiProvider(ai.config, ai.SpecialFunc)
+	if gemini != nil {
+		ai.provider["gemini"] = gemini
+	}
+	anthropic := providers.NewAnthropicProvider(ai.config, ai.SpecialFunc)
+	if anthropic != nil {
+		ai.provider["anthropic"] = anthropic
+	}
```
This looks rather repetitive to me. It should probably be handled in somekinda loop.

abenz1267 replied:
Also: do you really need the "New*Provider" functions? To me it looks like everything should be handled in the providers `SetupData` method, since you have a loop calling that already anyways. in `ai.SetupData`.

reflog replied:
@abenz1267 New*Provider functions are needed to validate env key presence and to remove the provider if it's not set up.

I did clean up the initialization using a map of known providers
